cpp/src/community/leiden_kernels.cu

```diff
@@ -0,0 +1,302 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <graph.hpp>
+
+#include <rmm/thrust_rmm_allocator.h>
+
+#include <community/louvain_kernels.hpp>
+#include <utilities/cuda_utils.cuh>
+#include <utilities/graph_utils.cuh>
+
+//#define TIMING
+
+#ifdef TIMING
+#include <utilities/high_res_timer.hpp>
+#endif
+
+#include <converters/COOtoCSR.cuh>
+
+namespace cugraph {
+namespace detail {
+
+template <typename vertex_t, typename edge_t, typename weight_t>
+weight_t update_clustering_by_delta_modularity_constrained(
+  weight_t total_edge_weight,
+  weight_t resolution,
+  GraphCSRView<vertex_t, edge_t, weight_t> const &graph,
+  rmm::device_vector<vertex_t> const &src_indices,
+  rmm::device_vector<weight_t> const &vertex_weights,
+  rmm::device_vector<weight_t> &cluster_weights,
+  rmm::device_vector<vertex_t> &cluster,
+  rmm::device_vector<vertex_t> &constraint,
+  cudaStream_t stream)
+{
+  rmm::device_vector<vertex_t> next_cluster(cluster);
+  rmm::device_vector<weight_t> delta_Q(graph.number_of_edges);
+  rmm::device_vector<vertex_t> cluster_hash(graph.number_of_edges);
+  rmm::device_vector<weight_t> old_cluster_sum(graph.number_of_vertices);
+
+  vertex_t *d_cluster_hash         = cluster_hash.data().get();
+  vertex_t *d_cluster              = cluster.data().get();
+  weight_t const *d_vertex_weights = vertex_weights.data().get();
+  weight_t *d_cluster_weights      = cluster_weights.data().get();
+  weight_t *d_delta_Q              = delta_Q.data().get();
+  vertex_t *d_constraint           = constraint.data().get();
+  vertex_t const *d_src_indices    = src_indices.data().get();
+  vertex_t const *d_dst_indices    = graph.indices;
+
+  weight_t new_Q = modularity<vertex_t, edge_t, weight_t>(
```

Comment:
Explicit template arguments should _not_ be necessary. The compiler should be able to deduce the `vertex_t, edge_t, weight_t` arguments from the `graph` argument. Explicit template arguments defeat the purpose of static polymorphism (via templates).

Reply:
Example below should compile and link just fine:
```
#include <iostream>

using cudaStream_t = int;

template<typename vertex_t, typename edge_t, typename weight_t>
struct GraphCSRView
{
};

template <typename vertex_t, typename edge_t, typename weight_t>
weight_t modularity(weight_t total_edge_weight,
                    weight_t resolution,
                    GraphCSRView<vertex_t, edge_t, weight_t> const &graph,
                    vertex_t const *d_cluster,
                    cudaStream_t stream = 0)
{
    std::cout<<"modularity...\n";
    return resolution;
}

template <typename vertex_t, typename edge_t, typename weight_t>
void foo(GraphCSRView<vertex_t, edge_t, weight_t> const &graph)
{
    vertex_t* d_cluster{nullptr};
    modularity(1.0, 2.0, graph, d_cluster); // compiler deduces the template args from the `graph` argument
}

int main(void)
{
    GraphCSRView<int, int, double> graph;
    foo(graph);

    return 0;
}
```

Reply:
Updated to not have explicit template arguments.